experimental/InvariantTheory/src/InvariantTheory.jl

```diff
@@ -0,0 +1,666 @@
+        #check char(fld)
+        if sym == :SL 
+            R, _ = polynomial_ring(fld, :z => (1:m,1:m))
+            return ReductiveGroup(sym, m, R)
+        elseif sym == :torus
+            G = new()
+            G.group = (sym, m)
+            characteristic(fld) == 0 || error("Characteristic should be 0 for linearly reductive groups")
+            G.field = fld
+            return G
+        end
+    end
+
+    function ReductiveGroup(sym::Symbol, m::Int, pring::MPolyRing) #the ring input is the group ring
+        #check char(field)
+        G = new()
+        if sym != :SL
+            error("Only implemented for SLm")
+        end
+        fld = base_ring(pring)
+        characteristic(fld) == 0 || error("Characteristic should be 0 for linearly reductive groups")
+        G.field = fld
+        @assert m^2  == ngens(pring)
+        G.group = (sym,m)
+        G.reynolds_operator = reynolds_slm
+        M = matrix(pring, m, m, gens(pring))
+        G.canonical_representation = M
+        G.group_ideal = ideal([det(M) - 1])
+        #base ring of M has to be the same as the representation matrix when that is created later.
+        return G
+    end
+end
+
+function Base.show(io::IO, G::ReductiveGroup)
+    io = AbstractAlgebra.pretty(io)
+    if G.group[1] == :SL
+        print(io, "Reductive group ", G.group[1], G.group[2])
+    else
+        println(io, "Torus of degree ", group(G)[2])
+        print(IOContext(io, :supercompact => true), AbstractAlgebra.Indent(), "over the field ", field(G))
+        print(io, AbstractAlgebra.Dedent())
+    end
+end
+
+reductive_group(sym::Symbol, m::Int, R::MPolyRing) = ReductiveGroup(sym,m,R)
+reductive_group(sym::Symbol, m::Int, F::Field) =  ReductiveGroup(sym,m,F)
+group(G::ReductiveGroup) = G.group
+field(G::ReductiveGroup) = G.field
+reynolds_operator(G::ReductiveGroup) = G.reynolds_operator
+group_ideal(G::ReductiveGroup) = G.group_ideal
+canonical_representation(G::ReductiveGroup) = G.canonical_representation
+natural_representation(G::ReductiveGroup) = G.canonical_representation
+
+#####################
+#Representation of Reductive Groups. Embeds SLm in GLn. 
+#####################
+
+mutable struct RepresentationReductiveGroup
+    group::ReductiveGroup
+    rep_mat::AbstractAlgebra.Generic.MatSpaceElem
+    sym_deg::Tuple{Bool, Int}
+    reynolds_v::Function
+
+    weights::Vector{Vector{ZZRingElem}}
+    
+    #representation of group G over symmetric degree d
+    function RepresentationReductiveGroup(G::ReductiveGroup, d::Int)
+        R = new()
+        R.group = G
+        R.rep_mat = rep_mat_(G, d)
+        R.sym_deg = (true, d)
+        R.reynolds_v = reynolds_v_slm
+        return R
+    end
+
+    #function RepresentationReductiveGroup(G::ReductiveGroup, W::Union{ZZMatrix, Matrix{<:Integer}, Vector{<:Integer}})
+    function RepresentationReductiveGroup(G::ReductiveGroup, W::Vector{Vector{ZZRingElem}})
+        if !(G.group[1] == :torus)
+            return nothing
```

Comment:
It is very unusual for a constructor to return something of a different type; it would be more typical to throw an exception here. Or is there a reason not to do so?